Auto register Models

cboelter opened the pull request:
As discussed in #15 I have created a PR for this. Please check the PR an give Feedback if something is missing ...

dmolineus commented on src/Bundle/DependencyInjection/Compiler/RegisterContaoModelPass.php:
```diff
@@ -0,0 +1,85 @@
+     * @param string $tagName       Name of the tag.
+     * @param int    $argumentIndex Index of the argument which should get the tagged references.
+     */
+    public function __construct(string $tagName, int $argumentIndex)
```
I don't see any reuse case here, so I would hardcode the tag name and the index in this class.

dmolineus commented on src/Bundle/DependencyInjection/Compiler/RegisterContaoModelPass.php:
```diff
@@ -0,0 +1,85 @@
+                Assert::that($tag)->keyExists('model');
+                Assert::that($tag['model'])->string();
+                Assert::that($tag['model'])->classExists();
+                Assert::that($tag['model'])->subclassOf(Model::class);
```
`Assert::that` provides a fluent interface. So you can do:

```php
Assert::that($tag['model'])
    ->string()
    ->classExists()
    ->subclassOf(Model::class);
```

dmolineus commented on src/Bundle/Resources/config/listeners.yml:
```diff
@@ -39,6 +39,14 @@ services:
+    arguments:
+      - []
+    tags:
+      - { name: 'contao.hook', hook: 'initializeSystem', method: 'onInitializeSystem', priority: 101 }
```
Any reason why you set priority to `101`? I would use `100` here or do you see any issues?

cboelter replied:
No reason for that ... only a wrong number :-) Will change it!

cboelter commented:
@dmolineus Just for the Listener or also for the CompilerPass?

dmolineus commented:
> @dmolineus Just for the Listener or also for the CompilerPass?

For both please. Didn't recognizes the second is also missing. :smile:

cboelter commented:
@dmolineus I have changed the things you commented and added also a spec for the RegisterContaoModelPass, but I'm not really sure it is the right way, because I have never used that.

Can you check it please and make some suggestions? I'm really sure that the dummy data is wrong with the ContentModel :-(

dmolineus commented on spec/Bundle/DependencyInjection/Compiler/RegisterContaoModelPassSpec.php:
```diff
@@ -0,0 +1,66 @@
+                ['model' => ContentModel::class]
+            ],
+            'bar' => [
+                ['model' => ContentModel::class]
```
Any reasons testing with two repositories registering `ContentModel`? Actually this should throw an exception because you can't foresee which is the repository later on. You don't have to implement the exception catching, but  only use another model class here.

dmolineus commented on spec/Bundle/DependencyInjection/Compiler/RegisterContaoModelPassSpec.php:
```diff
@@ -0,0 +1,66 @@
+            ->shouldBeCalled()
+            ->willReturn($taggedServices);
+
+        $definition->replaceArgument(0, Argument::size(1))->shouldBeCalled();
```
I'd prefer testing against real values here. Instead of `Argument::size(1)` I would use `['tl_content' => ContentModel::class, ....]` here.

cboelter commented:
@dmolineus I have startet to work on the final requests. But I'm not really sure about the ```RegisterContaoModelsListenerSpec```. Do you have any example for me, where I can see how I have to build this spec?

dmolineus commented:
I already gave the hint in https://github.com/netzmacht/contao-toolkit/pull/16#pullrequestreview-217168013 ;-). Nevermind, I just added the spec to the PR.

PHPSpec isn't actually designed for checking global state. The helper library overcomes this limitation so that it's able to check such Contao specific stuff.

cboelter commented:
@dmolineus I've added my changes you requested, please check again! Thanks for the spec, never worked with it before.

cboelter commented:
If we are done it would be nice, if you can release it today :-) ... so i can release my depending extensions.

dmolineus commented:
LGTM. Thanks for the contribution. Will be part of the upcoming [3.3.0](https://github.com/netzmacht/contao-toolkit/milestone/6) release. There are still some issues left. I doubt there's no chance to get it released today. :(

dmolineus commented:
Just released version 3.3.0 containing this feature.
